docs/DafnyRef/Topics.md

```diff
@@ -10,6 +10,31 @@ TO BE WRITTEN
 
 TO BE WRITTEN
 
+### 23.2.1. Subset types
```

Comment:
The PR description contains this segment about when quantifiers over subset types with ghost constraints can still be compiled:

> Consider a comprehension like set c: GhostEvenCell | …, where the GhostEvenCell may not be run-time checkable. In this case, it may be impossible to emit compiled code. However, there are two cases where we can and will allow it:
- If the static type of the candidate values is already a subtype of GhostEvenCell, then everything is fine. For example, if the BoundedPool discovers c in S where S has type set<GhostEvenCell>, then everything is fine. Then, the compiler emits code that evaluates the body of the comprehension for every candidate value.
- NEW If the static type of the candidate values is a type T that is not a subtype of GhostEvenCell, then the verifier will check if, by chance, the body of the comprehension is written in such a way that (a) it is well-formed for every T value (as if bound variable c could have been declared to have type T) and (b) every T value that satisfies the body is in fact a GhostEvenCell. If the verifier is able to check this, then the comprehension is allowed and the compiler can use T as the static type of the bound variable used in the emitted code

Shouldn't that explanation be in the documentation as well?

Reply:
Let's be empirical: subset types are one of the things that are the most powerful things in Dafny, because they can embed pre and post conditions implicitly anywhere.

> Do users really need to compile code that references subset types with ghost constraints?
> What's the motivating example where a user needs to have a ghost constraint subset type be part of a quantifier?

Let me elaborate on this to explain to you why this is not only useful, but also necessary for real code.
The ghost constraint only indicates specifications that these values must adhere to, and one good thing is that they are not checked at compile-time, so it means performance improvement. But if they are not checked at compile-time, we need to verify that these constraint hold anyway.
It's like the "function by method" paradigm. We might want to ensure a predicate (the constraint of the subset type), but there might be a more efficient way to compute it.

For example, if you had:
```dafny
predicate P(t) { A && B }
predicate Q(t) { P(t) && C }
type TP := t | P(t)
type TQ := t | Q(t)
```
Let's say you have a `set<TP>` and you need a `set<TQ>` in a compiled context:
```dafny
var setOfTp: set<TP> := ...
set t: TQ | t in setOfTp
```
Without this PR, you'd have to 1) make the predicate `P` and `Q` not ghost, which transitively would result in every of their dependencies being not ghost, including constant fields. That's a bummer.
2) The predicate `Q(t)` would be checked for every element of `setOfTp`, which includes expressions we already know hold like `A` and `B`. That's a performance penalty that any user would prefer to avoid if possible.

With this PR, you don't change the ghostness of `P` or `Q`, and you'd write the set comprehension
```
set t: TQ | t in setOfTp && C
```
And that just works as expected, the verifier can prove after `&& C` that `t` is a `TQ`

> It seems simpler not to allow it. 
I'm not sure what you mean here. Of course it is simpler to not support a construct than to support it. Can you please elaborate a bit more?

> Can't the user just change the constraint of their subset type so it's no longer a ghost constraint?
First, it's not always possible. Some constraints might refer to ghost fields of classes for example. Second, as explained before, verification aims at removing run-time checking of constraints so that the code can be more performant.
So, if a user can do it, they can definitely change a constraint to be compilable and they won't have any problem, because that constraint will be checked. But this PR gives them the freedom to choose.

Reply:
Both of your last examples are not allowed, because `Q` is non-ghost and is trying to call a ghost predicate in an expression that is not a specification of `Q`
What you probably want is to compare:

```dafny
predicate P(t) { A && B }
predicate Q(t) { P(t) && C }
type TP := t | P(t)
type TQ := t | Q(t)
set t: TQ | t in setOfTp && C
```

with 

```dafny
predicate method P(t) { A && B }
predicate method Q(t) { P(t) && C }
type TP := t | P(t)
type TQ := t | Q(t)
set t: TQ | t in setOfTp
```

The first one will just enumerate all the elements of `setOfTp` and then check if `C` holds, whereas the second one having not necessarily compatible types will need to check `Q(t)` for every `t`.

However, there is still an open question for this one:

```dafny
predicate method P(t) { A && B }
predicate method Q(t) { P(t) && C }
type TP := t | P(t)
type TQ := t | Q(t)
set t: TQ | t in setOfTp && C
```

Here, theoretically, the verifier should be able to prove that it's not necessary to do the check `Q(t)` for every element since `C` enforce it.
However, this mechanism might be very hard to enforce. Currently, because when the predicates are compilable, the type of `t` in `t in setOfTp && C` is inferred by the resolver to be `TQ` directly. The reason for that is that this condition can be checked by the compiler, so it's safe to do so, so C will hold anyway.
If someone would really like to have compiled predicates but benefit from an optimization of not re-checking the previous predicates, in this context, they would be able to do so by writing the following:

 ```dafny
predicate method P(t) { A && B }
predicate method Q(t) { P(t) && C }
type TP := t | P(t)
type TQ := t | Q(t)
set t: TQ, p: TP | p in setOfTp && C[t/p] && t == p :: t
```
This is saying "Take all the elements of `setOfTp` with their original type, verify the additional constraint, and if it holds assign it to t, thus verifying that this constraint hold", which I believe is acceptable.

Reply:
> Both of your last examples are not allowed, because `Q` is non-ghost and is trying to call a ghost predicate in an expression that is not a specification of `Q` What you probably want is to compare:
> 
> ```
> predicate P(t) { A && B }
> predicate Q(t) { P(t) && C }
> type TP := t | P(t)
> type TQ := t | Q(t)
> set t: TQ | t in setOfTp && C
> ```
> 
> with
> 
> ```
> predicate method P(t) { A && B }
> predicate method Q(t) { P(t) && C }
> type TP := t | P(t)
> type TQ := t | Q(t)
> set t: TQ | t in setOfTp
> ```
> 
> The first one will just enumerate all the elements of `setOfTp` and then check if `C` holds, whereas the second one having not necessarily compatible types will need to check `Q(t)` for every `t`.
> 
> However, there is still an open question for this one:
> 
> ```
> predicate method P(t) { A && B }
> predicate method Q(t) { P(t) && C }
> type TP := t | P(t)
> type TQ := t | Q(t)
> set t: TQ | t in setOfTp && C
> ```
> 
> Here, theoretically, the verifier should be able to prove that it's not necessary to do the check `Q(t)` for every element since `C` enforce it. However, this mechanism might be very hard to enforce. Currently, because when the predicates are compilable, the type of `t` in `t in setOfTp && C` is inferred by the resolver to be `TQ` directly. The reason for that is that this condition can be checked by the compiler, so it's safe to do so, so C will hold anyway. If someone would really like to have compiled predicates but benefit from an optimization of not re-checking the previous predicates, in this context, they would be able to do so by writing the following:
> 
> ```
> predicate method P(t) { A && B }
> predicate method Q(t) { P(t) && C }
> type TP := t | P(t)
> type TQ := t | Q(t)
> set t: TQ, p: TP | p in setOfTp && C[t/p] && t == p :: t
> ```
> 
> This is saying "Take all the elements of `setOfTp` with their original type, verify the additional constraint, and if it holds assign it to t, thus verifying that this constraint hold", which I believe is acceptable.

Thanks for the great explanation, and you were right about what I probably wanted.

Can we add the various alternatives and their performance impact to the documentation? Please correct me because I'm probably wrong, but these are the alternatives I see now:
- Compilable constraint subset type that is checked at runtime. (this one is already in the documentation)
- Compilable constraint subset type that is checked at compile-time. (your last example)
- Ghost constraint subset type that is not allowed (this one is already in the documentation)
- Ghost constraint subset type that is allowed because
  - Source collection is already a subset of the bound variable type
  - Comprehension range constraint specifies a subset of the bound variable type

Reply:
One more thing, is the emitting of runtime checked constraints something that is added in this PR, or did that already exist?

My preference would be for runtime checks to require an explicit piece of syntax. Something like `t is TQ` in the following:

```
predicate method P(t) { A && B }
predicate method Q(t) { P(t) && C }
type TP := t | P(t)
type TQ := t | Q(t)
set t: TQ | t in setOfTp && t is TQ
```

Reply:
> One more thing, is the emitting of runtime checked constraints something that is added in this PR, or did that already exist?

That emission already existed before, in the previous `CreateForEach` method, when the subset type had a compilable constraint:
https://github.com/dafny-lang/dafny/pull/1702/files#diff-daf85b4917468b5fae254d3b53259f20e3558d36340819d1e91a2b6ba2d4e18fL4663
https://github.com/dafny-lang/dafny/pull/1702/files#diff-daf85b4917468b5fae254d3b53259f20e3558d36340819d1e91a2b6ba2d4e18fL4705
It was was not always checking the subtype as well (hence the compiler soundness bug).

Your point is a very interest one and it might require more consideration, because it would break existing code, which i tried to avoid if possible. For example:

```dafny
var s := set x: nat | x in setOfInt
```
verifies and compiles, before this PR. This PR keeps that same behavior.

Your idea needs the check `x >= 0` or `x is nat` needs to become explicit. We might want to keep this change for a big version change as a separate PR, if we agree on that (which is possible, after all, I like performance and I like it when things are not made invisibly).

Reply:
> Can we add the various alternatives and their performance impact to the documentation? Please correct me if I'm wrong, but these are the alternatives I see now:
> 
> * Compilable constraint subset type that is checked at runtime. (this one is already in the documentation)
> * Compilable constraint subset type that is checked at compile-time. (your last example)
> * Ghost constraint subset type that is not allowed (this one is already in the documentation)
> * Ghost constraint subset type that is allowed because
>   
>   * Source collection is already a subset of the bound variable type
>   * Comprehension range constraint specifies a subset of the bound variable type

I added three examples to cover these missing cases in the documentation.